src/type-data-frame.c

```diff
@@ -241,3 +245,64 @@ SEXP df_poke_at(SEXP x, SEXP name, SEXP value) {
   UNPROTECT(1);
   return x;
 }
+
+// [[ include("type-data-frame.h") ]]
+R_len_t df_flat_width(SEXP x) {
+  R_len_t n = Rf_length(x);
+  R_len_t out = n;
+
+  for (R_len_t i = 0; i < n; ++i) {
+    SEXP col = VECTOR_ELT(x, i);
+    if (is_data_frame(col)) {
+      out = out + df_flat_width(col) - 1;
+    }
+  }
+
+  return out;
+}
+
+// [[ register() ]]
+SEXP vctrs_df_flat_width(SEXP x) {
+  return r_int(df_flat_width(x));
+}
+
+
+static R_len_t df_flatten_loop(SEXP x, SEXP out, SEXP out_names, R_len_t counter);
+
+// Might return duplicate names. Currently only used for equality
+// proxy so this doesn't matter. A less bare bone version would repair
+// names.
+//
+// [[ register(); include("type-data-frame.h") ]]
+SEXP df_flatten(SEXP x) {
+  R_len_t width = df_flat_width(x);
+  SEXP out = PROTECT(Rf_allocVector(VECSXP, width));
+  SEXP out_names = PROTECT(Rf_allocVector(STRSXP, width));
+  r_poke_names(out, out_names);
+
+  df_flatten_loop(x, out, out_names, 0);
+  init_data_frame(out, df_size(x));
+
+  UNPROTECT(2);
+  return out;
+}
+
+static R_len_t df_flatten_loop(SEXP x, SEXP out, SEXP out_names, R_len_t counter) {
+  R_len_t n = Rf_length(x);
+  SEXP x_names = PROTECT(r_names(x));
```

Comment:
Do we need to worry about the very rare case of a data frame with no column names? You call `STRING_ELT()` on the names and they could be `NULL` in that case

Reply:
I think this would be a case of a corrupt data structure? In that case the `string-elt` error about wrong type-of seems sufficiently defensive.